### PR TITLE
Add BaseModel class to help auto generate attribute accessors

### PR DESCRIPTION
**Pull request by bzwei**
The `BaseModel` class is borrowed from our armrest gem. It automatically converts all hash attributes to instance methods. It can preserve named list of hash attribute from being converted. So far I don't know which attributes need to remain as hash.

With this work module `InstanceMethods` is no longer needed.

Update: attributes that appear in the related field will be treated as ids. So job_template, inventory, cloud_credential, project etc will be automatically appended with _id.

### CODE REVIEW

**Comment by gmcculloug** [comment]
@bdunne @syncrou Please review.

File: lib/ansible_tower_client/group.rb [hunk 1]
```diff
@@ -1,14 +1,6 @@
 module AnsibleTowerClient
-  class Group
+  class Group < BaseModel
     extend CollectionMethods
-    include InstanceMethods
-
-    attr_reader :inventory_id
-
-    def initialize(raw_body)
-      @inventory_id = raw_body["inventory"]
```
Comment (syncrou) [inline lib/ansible_tower_client/group.rb]: How do we implement a mapping back to \@inventory_id?

Comment (bzwei) [inline lib/ansible_tower_client/group.rb]: You can define your own method for particular model, for example
```
def inventory_id
  __getobj__['inventory']
end
```
The auto generated `inventory` method still exists and returns the same thing.

**Comment by syncrou** [comment]
:+1:  LGTM - Except in regards to the removed inventory_id attribute for Group, Host - I thought the app was expecting a return on that method. If not - I'm set.

**Comment by bzwei** [comment]
@syncrou If `inventory_id` is already used in manageiq app, then we need to update manageiq also. Or we can add back `inventory_id` method.

**Comment by bdunne** [comment]
@bzwei Should we extract [MiqHashStruct](https://github.com/ManageIQ/manageiq/blob/master/gems/pending/util/miq-hash_struct.rb) to more_core_extensions and use it as a superclass for BaseModel?  I think that would simplify the code in BaseModel.

**Comment by bdunne** [comment]
@bzwei I'm still stuck on `AnsibleTowerClient::Host#inventory.class` returning `Integer` rather than `AnsibleTowerClient::Inventory`.

**Comment by bzwei** [comment]
@bdunne @syncrou Update: attributes that appear in the `related` field will be treated as ids. So `job_template`, `inventory`, `cloud_credential`, `project` etc will be automatically appended with `_id`.

**Comment by bzwei** [comment]
@bdunne @syncrou please review and merge.

**Comment by syncrou** [comment]
@bzwei - :+1:

File: lib/ansible_tower_client/base_model.rb [hunk 2]
```diff
@@ -1,9 +1,152 @@
...
+    #   person = Person.new(json_string)
+    #
+    #   # The JSON properties are now available as methods.
+    #   person.firstname        # => 'jeff'
+    #   person.address.zipcode  # => '01013'
+    #
+    #   # Or you can get back the original JSON if necessary.
+    #   person.to_json # => Returns original JSON
+    #
+    def initialize(api, json = {})
```
Comment (bdunne) [inline lib/ansible_tower_client/base_model.rb]: Why is `json` optional?

File: lib/ansible_tower_client/base_model.rb [hunk 3]
```diff
@@ -1,9 +1,152 @@
...
+      # '#' is the separator between levels. Remove attributes
+      # before the first separator.
+      @child_excl_list = self.class.send(:excl_list).map do |e|
+        e.index('#') ? e[e.index('#') + 1..-1] : ''
+      end
+
+      __setobj__(json.kind_of?(Hash) ? json.dup : JSON.parse(json))
+    end
+
+    def to_h
```
Comment (bdunne) [inline lib/ansible_tower_client/base_model.rb]: Should this be an alias?

File: lib/ansible_tower_client/base_model.rb [hunk 4]
```diff
@@ -1,9 +1,152 @@
...
+
+    def to_hash
+      __getobj__.dup
+    end
+
+    def to_json
+      __getobj__.to_json
+    end
+
+    def to_s
```
Comment (bdunne) [inline lib/ansible_tower_client/base_model.rb]: Should this be an alias?

File: lib/ansible_tower_client/base_model.rb [hunk 5]
```diff
@@ -1,9 +1,152 @@
...
+        child_excl_list = @child_excl_list
+        self.class.const_set(klass_name, Class.new(BaseModel) { attr_hash(*child_excl_list) })
+      end
+      self.class.const_get(klass_name).new(api, value)
+    end
+
+    def add_accessor_methods(method, key)
+      method.prepend('_') if methods.include?(method.to_sym)
+      instance_eval { define_singleton_method(method) { __getobj__[key] } }
+      instance_eval { define_singleton_method("#{method}=") { |val| __getobj__[key] = val } }
```
Comment (bdunne) [inline lib/ansible_tower_client/base_model.rb]: If i'm reading this right, we will define `"#{method}=` for all attributes that will write to the objects underlying `@hashobj`, but not make any calls to the REST API.  I would expect `"#{method}=` to call the REST API to update an attribute then probably refresh the underlying `hashobj`, I don't think we should dynamically define all of those methods.

File: lib/ansible_tower_client/base_model.rb [hunk 6]
```diff
@@ -1,9 +1,152 @@
...
+      return false unless other.kind_of?(BaseModel)
+      __getobj__.eql?(other.__getobj__)
+    end
+
+    # Support hash style accessors
+    def [](key)
+      __getobj__[key]
+    end
+
+    def []=(key, val)
```
Comment (bdunne) [inline lib/ansible_tower_client/base_model.rb]: Same concern as https://github.com/ManageIQ/ansible_tower_client/pull/17/files#r55891091

File: spec/support/factory_helper.rb [hunk 7]
```diff
@@ -2,8 +2,8 @@
 
 module AnsibleTowerClient
   module FactoryHelper
-    KEYS = [:description, :extra_vars, :id, :instance_id,
-            :inventory, :name, :results, :type, :url, :related].freeze
+    KEYS = [:related, :description, :extra_vars, :id, :instance_id,
```
Comment (bdunne) [inline spec/support/factory_helper.rb]: Is there a reason this can't be in alphabetical order?

File: lib/ansible_tower_client/base_model.rb [hunk 8]
```diff
@@ -1,9 +1,152 @@
...
+      self.class.const_get(klass_name).new(api, value)
+    end
+
+    def add_accessor_methods(method, key)
+      method.prepend('_') if methods.include?(method.to_sym)
+      instance_eval { define_singleton_method(method) { __getobj__[key] } }
+      instance_eval { define_singleton_method("#{method}=") { |val| __getobj__[key] = val } }
+    end
+
+    def snake_case(name)
```
Comment (bdunne) [inline lib/ansible_tower_client/base_model.rb]: With ActiveSupport, you should be able to remove this in favor of `"".underscore`.

**Comment by bdunne** [comment]
@bzwei I think this PR is getting much closer.  I still think we should split `BaseModel` into:
- `BaseModel` with an initialize method that filters what attribute methods should be defined and
- its underlying `ReadOnlyHashStruct`

Does anyone else have an opinion? @Fryguy @gmcculloug @syncrou

**Comment by bdunne** [comment]
Also, If BaseModel is split, the ReadOnlyHashStruct can do something very similar to MiqHashStruct using `method_missing` and `respond_to_missing?` to hit`:[]`.  I think that will simplify the code a lot and remove the need for the `instance_eval` and `define_method` lines.

File: lib/ansible_tower_client/base_model.rb (hunk 5 above)
Comment (bzwei) [inline lib/ansible_tower_client/base_model.rb]: The model class when used in armrest gem does not contain the service class/provider connection. It services purely as an attributes container. Therefore it makes sense to support both read and write.

In this gem, if you expect any changes in the model to be in sync with its corresponding provider object, then we should not auto_generate the setter.

File: spec/support/factory_helper.rb (hunk 7 above)
Comment (bzwei) [inline spec/support/factory_helper.rb]: `:related` has to be the first because its entries decide what attributes needs to be appended with `_id`

**Comment by bzwei** [comment]
@bdunne the nice thing with this implementation is that it is more like a real class. You can call #methods(false) to get all new methods it added based on the hash.

File: spec/support/factory_helper.rb (hunk 7 above)
Comment (bdunne) [inline spec/support/factory_helper.rb]: I wouldn't rely on it being first in the response.

**Pull request merged**